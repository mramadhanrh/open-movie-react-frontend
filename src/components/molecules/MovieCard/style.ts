import styled from 'styled-components';

export const Card = styled.button`
  border: none;
  background: #ffffff;
  padding: 0;
  margin: 10px 0;
  width: 100%;
  min-height: 20px;
  box-shadow: 0px 2px 20px rgba(0, 0, 0, 0.16);
  border-radius: 10px;
  max-width: 300px;
  text-align: left;
  overflow: hidden;
`;

export const CardImage = styled.img<{ src?: string }>`
  width: 100%;
`;

export const InfoWrapper = styled.div`
  padding: 10px 10px 15px;

  & > h1 {
    font-size: 24px;
    margin: 0;
  }
`;

export const BadgesWrapper = styled.div`
  display: flex;
  text-transform: capitalize;
  & > div {
    margin-right: 5px;
  }
`;
