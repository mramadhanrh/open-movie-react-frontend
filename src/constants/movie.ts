export const MovieEmptyPicture = {
  label: 'N/A',
  src: 'https://images.unsplash.com/photo-1578328819058-b69f3a3b0f6b?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=967&q=80',
};
