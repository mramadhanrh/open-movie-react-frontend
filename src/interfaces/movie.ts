export enum MovieResultType {
  Movie = 'movie',
  Series = 'series',
  Episode = 'episode',
}

export interface MovieItem {
  Title: string;
  Year: string;
  imdbID: string;
  Type: MovieResultType;
  Poster: string;
}

export interface MovieSearchResponse {
  Search: MovieItem[];
  totalResults: string;
  Response: 'True' | 'False';
  Error?: string;
}

export interface MovieRating {
  Source: string;
  Value: string;
}

export interface MovieDetailResponse {
  Title: string;
  Year: string;
  Rated: string;
  Released: string;
  Runtime: string;
  Genre: string;
  Director: string;
  Writer: string;
  Actors: string;
  Plot: string;
  Language: string;
  Country: string;
  Awards: string;
  Poster: string;
  Ratings: MovieRating[];
  Metascore: string;
  imdbRating: string;
  imdbVotes: string;
  imdbID: string;
  Type: MovieResultType;
  DVD: string;
  BoxOffice: string;
  Production: string;
  Website: string;
  Response: string;
  Error?: string;
}
