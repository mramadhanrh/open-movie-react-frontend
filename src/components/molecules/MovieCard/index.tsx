import { FC } from 'react';
import { MovieResultType } from '../../../interfaces/movie';
import Badges from '../../atoms/Badges';
import { BadgesWrapper, InfoWrapper, Card, CardImage } from './style';

interface MovieCardProps {
  title?: string;
  picture?: string;
  type?: MovieResultType;
  year?: string;
}

const MovieCard: FC<MovieCardProps> = ({ title, picture, type, year }) => (
  <Card>
    <CardImage src={picture} />
    <InfoWrapper>
      <h1>{title}</h1>
      <BadgesWrapper>
        <Badges color="primary">{type}</Badges>
        <Badges>{year}</Badges>
      </BadgesWrapper>
    </InfoWrapper>
  </Card>
);

export default MovieCard;
