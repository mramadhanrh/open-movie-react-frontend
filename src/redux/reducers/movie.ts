import { MovieItem } from '../../interfaces/movie';
import { MovieAction, MovieType } from '../types/movie';

interface MovieState {
  list: MovieItem[];
}

const initialState = {
  list: [],
};

const { SetMovieList, AddMovieList } = MovieType;

const movie = (
  state: MovieState = initialState,
  action: MovieAction,
): MovieState => {
  const { type, payload } = action;

  const setMovieList = (): MovieState => ({
    ...state,
    list: payload,
  });

  const addMovieList = (): MovieState => ({
    ...state,
    list: [...state.list, ...(payload || [])],
  });

  switch (type) {
    case SetMovieList:
      return setMovieList();
    case AddMovieList:
      return addMovieList();
    default:
      return state;
  }
};
export default movie;
