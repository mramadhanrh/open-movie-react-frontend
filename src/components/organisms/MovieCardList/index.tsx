import { FC } from 'react';
import { MovieEmptyPicture } from '../../../constants/movie';
import { MovieItem } from '../../../interfaces/movie';
import MovieCard from '../../molecules/MovieCard';
import { Container } from './style';

interface MovieCardListProps {
  data: MovieItem[];
}

const { label: emptyLabel, src: emptySrc } = MovieEmptyPicture;

const MovieCardList: FC<MovieCardListProps> = ({ data = [] }) => (
  <Container>
    {data.map(({ imdbID, Title, Poster, Year, Type }) => (
      <MovieCard
        key={imdbID}
        id={imdbID}
        title={Title}
        picture={Poster === emptyLabel ? emptySrc : Poster}
        type={Type}
        year={Year}
      />
    ))}
  </Container>
);
export default MovieCardList;
