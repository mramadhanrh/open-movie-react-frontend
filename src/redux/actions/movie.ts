import { AxiosResponse } from 'axios';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  MovieItem,
  MovieResultType,
  MovieSearchResponse,
} from '../../interfaces/movie';
import { getMovieList } from '../../services/searchMovie';
import { AppStateType } from '../reducers';
import { MovieAddList, MovieSetList, MovieType } from '../types/movie';

export const setMovieList = (list: MovieItem[]): MovieSetList => ({
  type: MovieType.SetMovieList,
  payload: list,
});

export const setAddMovieList = (list: MovieItem[]): MovieAddList => ({
  type: MovieType.AddMovieList,
  payload: list,
});

export const loadMovieInitList =
  (search: string, page = 1, type?: MovieResultType) =>
  async (
    dispatch: ThunkDispatch<AppStateType, null, AnyAction>,
  ): Promise<
    [MovieSearchResponse | null, AxiosResponse<MovieSearchResponse> | null]
  > => {
    try {
      const res = await getMovieList(search, page, type);
      dispatch(setMovieList(res.data.Search));

      return [res.data, null];
    } catch (e) {
      return [null, e];
    }
  };

export const loadMovieAddList =
  (search: string, page = 1, type?: MovieResultType) =>
  async (
    dispatch: ThunkDispatch<AppStateType, null, AnyAction>,
  ): Promise<
    [MovieSearchResponse | null, AxiosResponse<MovieSearchResponse> | null]
  > => {
    try {
      const res = await getMovieList(search, page, type);
      dispatch(setAddMovieList(res.data.Search));

      return [res.data, null];
    } catch (e) {
      return [null, e];
    }
  };
