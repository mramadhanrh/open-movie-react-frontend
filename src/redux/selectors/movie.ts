import { MovieDetailResponse, MovieItem } from '../../interfaces/movie';
import { AppStateType } from '../reducers';

export const movieListSelector = (state: AppStateType): MovieItem[] =>
  state.movie.list;

export const movieDetailSelector = (
  state: AppStateType,
  id: string,
): MovieDetailResponse => state.movie.detail[id];
