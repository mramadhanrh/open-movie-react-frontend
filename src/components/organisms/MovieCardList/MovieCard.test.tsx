import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MovieCardList from '.';
import { MovieItem, MovieResultType } from '../../../interfaces/movie';
import MovieCard from '../../molecules/MovieCard';

const mockData: MovieItem[] = [
  {
    Title: 'Captain Marvel',
    Year: '2019',
    imdbID: 'tt4154664',
    Type: MovieResultType.Movie,
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMTE0YWFmOTMtYTU2ZS00ZTIxLWE3OTEtYTNiYzBkZjViZThiXkEyXkFqcGdeQXVyODMzMzQ4OTI@._V1_SX300.jpg',
  },
  {
    Title: 'Marvel One-Shot: Agent Carter',
    Year: '2013',
    imdbID: 'tt3067038',
    Type: MovieResultType.Movie,
    Poster:
      'https://m.media-amazon.com/images/M/MV5BZDIwZTM4M2QtMWFhYy00N2VmLWFlMjItMzI3NjBjYTc0OTMxXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg',
  },
  {
    Title: 'Marvel One-Shot: All Hail the King',
    Year: '2014',
    imdbID: 'tt3438640',
    Type: MovieResultType.Movie,
    Poster:
      'https://m.media-amazon.com/images/M/MV5BZGFkMTZkMDQtNzM4Yy00YWEwLTkzOWEtZTMyNDRlNmJhYWJhXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg',
  },
  {
    Title: 'Marvel One-Shot: Item 47',
    Year: '2012',
    imdbID: 'tt2247732',
    Type: MovieResultType.Movie,
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMjNlMzAxNmQtOGEwZi00NTEyLWI0NWYtMTlhNmE2YTA3ZDVhXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg',
  },
  {
    Title:
      "Marvel One-Shot: A Funny Thing Happened on the Way to Thor's Hammer",
    Year: '2011',
    imdbID: 'tt2011109',
    Type: MovieResultType.Movie,
    Poster:
      'https://m.media-amazon.com/images/M/MV5BYmVlYTg3N2QtMWM2OS00YWQyLWI2M2MtMDc0ZjBkZjk1MTY3XkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg',
  },
];

describe('MovieCardList', () => {
  const listComp = <MovieCardList data={mockData} />;
  test('match snapshot', () => {
    expect(toJson(shallow(listComp))).toMatchSnapshot();
  });

  test('count match with data', () => {
    expect(shallow(listComp).find(MovieCard).length).toBe(mockData.length);
  });
});
