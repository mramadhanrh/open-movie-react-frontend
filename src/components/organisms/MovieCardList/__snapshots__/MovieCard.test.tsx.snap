// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCardList match snapshot 1`] = `
<styled.div>
  <MovieCard
    id="tt4154664"
    key="tt4154664"
    onClick={[Function]}
    picture="https://m.media-amazon.com/images/M/MV5BMTE0YWFmOTMtYTU2ZS00ZTIxLWE3OTEtYTNiYzBkZjViZThiXkEyXkFqcGdeQXVyODMzMzQ4OTI@._V1_SX300.jpg"
    title="Captain Marvel"
    type="movie"
    year="2019"
  />
  <MovieCard
    id="tt3067038"
    key="tt3067038"
    onClick={[Function]}
    picture="https://m.media-amazon.com/images/M/MV5BZDIwZTM4M2QtMWFhYy00N2VmLWFlMjItMzI3NjBjYTc0OTMxXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg"
    title="Marvel One-Shot: Agent Carter"
    type="movie"
    year="2013"
  />
  <MovieCard
    id="tt3438640"
    key="tt3438640"
    onClick={[Function]}
    picture="https://m.media-amazon.com/images/M/MV5BZGFkMTZkMDQtNzM4Yy00YWEwLTkzOWEtZTMyNDRlNmJhYWJhXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg"
    title="Marvel One-Shot: All Hail the King"
    type="movie"
    year="2014"
  />
  <MovieCard
    id="tt2247732"
    key="tt2247732"
    onClick={[Function]}
    picture="https://m.media-amazon.com/images/M/MV5BMjNlMzAxNmQtOGEwZi00NTEyLWI0NWYtMTlhNmE2YTA3ZDVhXkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg"
    title="Marvel One-Shot: Item 47"
    type="movie"
    year="2012"
  />
  <MovieCard
    id="tt2011109"
    key="tt2011109"
    onClick={[Function]}
    picture="https://m.media-amazon.com/images/M/MV5BYmVlYTg3N2QtMWM2OS00YWQyLWI2M2MtMDc0ZjBkZjk1MTY3XkEyXkFqcGdeQXVyNTE1NjY5Mg@@._V1_SX300.jpg"
    title="Marvel One-Shot: A Funny Thing Happened on the Way to Thor's Hammer"
    type="movie"
    year="2011"
  />
</styled.div>
`;
